### PR TITLE
Recipe Recommender

### PR DESCRIPTION
**Pull request by Armintron**
Scrapes http://whatthefuckshouldimakefordinner.com/ to get a random recipe, providing a preview image and a link to the recipe

### CODE REVIEW

**Comment by suneettipirneni** [comment]
This needs a pull from develop

**Comment by tanndlin** [comment]
> This needs a pull from develop

prob just because of wrong base

**Comment by suneettipirneni** [comment]
> > This needs a pull from develop
> 
> 
> 
> prob just because of wrong base

Well *and* that this develop branch is out of date

**Comment by suneettipirneni** [comment]
I think the commit history here is messed up, it's giving diffs for things already merged

**Comment by tanndlin** [comment]
idk what you just did, it was fine before you did that

**Comment by tanndlin** [comment]
Want me to force push a fixed branch?

**Comment by Armintron** [comment]
Please 😔

**Comment by tanndlin** [comment]
Still need to use the inherited property, I didn't do that for you. All i did was remove the unnecessary commits and rebase on develop

File: src/app/plugins/reciperecommender.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  public usage: string = 'recrec <veg>?';
+  public permission: ChannelType = ChannelType.Public;
+  public pluginAlias: string[] = ['recrec', 'rr', 'reciperecommender'];
+  public pluginChannelName = Constants.Channels.Public.Food;
+
+  private readonly _API_URL = 'http://whatthefuckshouldimakefordinner.com/';
+  private readonly _RECIPE_NAME_REGEX = /\>(.*?)\</g;
+  private readonly _IMAGE_PREVIEW_REGEX = /<meta.*property="og:image".*content="(.*)".*\/>/;
+  private readonly _URL_REGEX = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&/=]*)/;
+  private readonly _AUTHOR_IMAGE_URL = 'https://i.imgur.com/B2vwvFL.png';
```
Comment (suneettipirneni) [inline src/app/plugins/reciperecommender.plugin.ts:21]: These should all be `static`

**Review by suneettipirneni — CHANGES_REQUESTED** [review]

File: src/app/plugins/reciperecommender.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    const [, previewImageURL] = (await axios.get(recipeURL)).data.match(RecipeRecommenderPlugin._IMAGE_PREVIEW_REGEX);
+    const [, uglyRecipeName] = scrapedRecipe.match(RecipeRecommenderPlugin._RECIPE_NAME_REGEX)!;
+    const recipeName = uglyRecipeName.substring(1, uglyRecipeName.length - 1);
+
+    const embed = new MessageEmbed();
+    embed
+      .setAuthor('What the f*** should I make today?', RecipeRecommenderPlugin._AUTHOR_IMAGE_URL, API_URL)
+      .setTitle(recipeName)
+      .setURL(recipeURL)
+      .setImage(previewImageURL ?? '');
```
Comment (suneettipirneni) [inline src/app/plugins/reciperecommender.plugin.ts:50]:
```suggestion
    const embed = new MessageEmbed()
      .setAuthor('What the f*** should I make today?', RecipeRecommenderPlugin._AUTHOR_IMAGE_URL, API_URL)
      .setTitle(recipeName)
      .setURL(recipeURL)
      .setImage(previewImageURL ?? '');
```

**Review by suneettipirneni — CHANGES_REQUESTED** [review]

**Review by suneettipirneni — APPROVED** [review]

File: src/app/plugins/reciperecommender.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    // response page is always the same so we just take a specific line from the HTML response
+    // based on wantsVeg, which changes the api endpoint used
+    const scrapedRecipe: string = res.data.split('\n')[wantsVeg ? 23 : 21];
+    const [recipeURL] = scrapedRecipe.match(RecipeRecommenderPlugin._URL_REGEX)!;
+    // scrapes <meta content="og:image"> from recipie website
+    const [, previewImageURL] = (await axios.get(recipeURL)).data.match(RecipeRecommenderPlugin._IMAGE_PREVIEW_REGEX);
+    const [, uglyRecipeName] = scrapedRecipe.match(RecipeRecommenderPlugin._RECIPE_NAME_REGEX)!;
+    const recipeName = uglyRecipeName.substring(1, uglyRecipeName.length - 1);
```
Comment (tanndlin) [inline src/app/plugins/reciperecommender.plugin.ts:43]: I don't really like any of this
I know its hard bc we are parsing HTML but ew

File: src/app/plugins/reciperecommender.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  private static readonly _IMAGE_PREVIEW_REGEX = /<meta.*property="og:image".*content="(.*)".*\/>/;
+  private static readonly _URL_REGEX = /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&/=]*)/;
+  private static readonly _AUTHOR_IMAGE_URL = 'https://i.imgur.com/B2vwvFL.png';
+
+  constructor(public container: IContainer) {
+    super();
+  }
+
+  public async execute(message: IMessage, args: string[]) {
+    const wantsVeg = args[0] === 'veg';
```
Comment (tanndlin) [inline src/app/plugins/reciperecommender.plugin.ts:28]: Does this not create an error when an argument is no given?

Also, should use `toLowerCase()`

File: src/app/plugins/reciperecommender.plugin.ts (hunk 3 above)
Comment (tanndlin) [reply src/app/plugins/reciperecommender.plugin.ts:43]: I feel like this portion should abstracted to a function that returns an interface for a recipe, bc its not readable as is

**Review by tanndlin — CHANGES_REQUESTED** [review]

File: src/app/plugins/reciperecommender.plugin.ts (hunk 4 above)
Comment (Armintron) [reply src/app/plugins/reciperecommender.plugin.ts:28]: Doesnt give an error, will make the lowercase change

File: src/app/plugins/reciperecommender.plugin.ts (hunk 3 above)
Comment (Armintron) [reply src/app/plugins/reciperecommender.plugin.ts:43]: > I don't really like any of this
> I know its hard bc we are parsing HTML but ew

Not really sure how else to do it.

Comment (suneettipirneni) [reply src/app/plugins/reciperecommender.plugin.ts:43]: You can use the `cheerio` library, it's already installed in lion.

Comment (tanndlin) [reply src/app/plugins/reciperecommender.plugin.ts:43]:
```ts
interface IRecipe{
    recipeURL: string;
    recipeName: string;
    previewImageURL: string;
}
```
and whatever else you need

**Comment by tanndlin** [comment]
this is very stale :(

**Pull request closed without merging**